### DISCUSSION TITLE
Loading anonymous javascript functions

### DISCUSSION THREAD

**Discussion by developer-uhwi** [Budibase/budibase#5893] [Help]
I want to be able to load the following function for (Chatwoot - customer engagement) in budibase using the embed component, this does not work. I want to add chat agent scripts in the application, to provide in app support, how can I do this. Also this would be helpful in the integration of other applications e.g. analytics like matomo or google.

 Running the function below in  the console currently gives me the following error.
`Refused to load the script 'https://base_url.com/packs/js/sdk.js' because it violates the following Content Security Policy directive: "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://cdn.budi.live https://js.intercomcdn.com https://widget.intercom.io". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.`

1. How do I load the anonymous function?
2. How do I add additional CSP headers or overwrite the current ones?
3. I am using nginx and have tried the following, this does not overwrite the CSPs already set in the application. 
`add_header Content-Security-Policy "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://cdn.budi.live https://js.intercomcdn.com https://widget.intercom.io https://base_url.com;" always;
`
See the function below.
` (function(d,t) {
        var BASE_URL="https://base_url.com";
        var g=d.createElement(t),s=d.getElementsByTagName(t)[0];
        g.src=BASE_URL+"/packs/js/sdk.js";
        g.defer = true;
        g.async = true;
        s.parentNode.insertBefore(g,s);
        g.onload=function(){
          window.chatwootSDK.run({
            websiteToken: 'xxxxxxx',
            baseUrl: BASE_URL
          })
        }
      })(document,"script");`

**Comment by melohagan** [comment]
Hi @developer-uhwi 

Do any of the answers here help? https://stackoverflow.com/questions/31211359/refused-to-load-the-script-because-it-violates-the-following-content-security-po

Do this at your own risk though; it could open you up to XSS.

**Comment by developer-uhwi** [comment]
I would rather not have XSS issues, just wondering how could I incorporate
integrations like these chat support tools like intercom or even Google
analytics.

**Reply by melohagan** [reply]
If you are self hosting, you could try editing the nginx config, but these changes will revert with updates: https://github.com/Budibase/budibase/blob/master/hosting/nginx.prod.conf.hbs

I'm curious what you're using Budibase for? Oftentimes Budibase is used for building internal business apps. Are you creating something customer facing?

**Reply by developer-uhwi** [reply]
This is really to provide in app help for internal users, while using applications.

**Reply by melohagan** [reply]
There's been a fair bit of change to Budibase since May! You may find that [custom components](https://docs.budibase.com/docs/custom-component) are useful for what you're trying to do here!

**Comment by developer-uhwi** [comment]
Will explore this
